File: services/account_snapshot.py
from __future__ import annotations

import json
import math
from pathlib import Path
from typing import Any, cast


class AccountSnapshotLimitError(ValueError):
    """The account snapshot exceeds its executable resource contract."""


_CONTRACT_PATH = Path(__file__).resolve().parent.parent / "account_snapshot_contract.json"
_CONTRACT_FIELDS = (
    "max_bytes",
    "max_records",
    "max_depth",
    "max_nodes",
    "max_object_fields",
    "max_key_bytes",
    "max_string_bytes",
)


def _load_contract() -> dict[str, int]:
    try:
        value = json.loads(_CONTRACT_PATH.read_text(encoding="utf-8"))
    except (OSError, UnicodeError, json.JSONDecodeError) as exc:
        raise RuntimeError("account snapshot contract is unavailable") from exc
    if (
        not isinstance(value, dict)
        or set(value) != set(_CONTRACT_FIELDS)
        or any(type(value.get(field)) is not int or value[field] <= 0 for field in _CONTRACT_FIELDS)
    ):
        raise RuntimeError("account snapshot contract is invalid")
    return cast(dict[str, int], value)


_CONTRACT = _load_contract()
ACCOUNT_SNAPSHOT_MAX_BYTES = _CONTRACT["max_bytes"]
ACCOUNT_SNAPSHOT_MAX_RECORDS = _CONTRACT["max_records"]
ACCOUNT_SNAPSHOT_MAX_DEPTH = _CONTRACT["max_depth"]
ACCOUNT_SNAPSHOT_MAX_NODES = _CONTRACT["max_nodes"]
ACCOUNT_SNAPSHOT_MAX_OBJECT_FIELDS = _CONTRACT["max_object_fields"]
ACCOUNT_SNAPSHOT_MAX_KEY_BYTES = _CONTRACT["max_key_bytes"]
ACCOUNT_SNAPSHOT_MAX_STRING_BYTES = _CONTRACT["max_string_bytes"]


class _JsonBudgetScanner:
    def __init__(self, payload: bytes) -> None:
        self.payload = payload
        self.length = len(payload)
        self.index = 0
        self.nodes = 0

    def scan(self) -> None:
        self._skip_whitespace()
        self._scan_value(1)
        self._skip_whitespace()
        if self.index != self.length:
            raise AccountSnapshotLimitError("invalid trailing JSON data")

    def _skip_whitespace(self) -> None:
        payload = self.payload
        index = self.index
        length = self.length
        while index < length and payload[index] in (0x20, 0x09, 0x0A, 0x0D):
            index += 1
        self.index = index

    def _consume(self, byte: int) -> None:
        if self.index >= self.length or self.payload[self.index] != byte:
            raise AccountSnapshotLimitError("invalid JSON syntax")
        self.index += 1

    def _count_node(self) -> None:
        self.nodes += 1
        if self.nodes > ACCOUNT_SNAPSHOT_MAX_NODES:
            raise AccountSnapshotLimitError("account snapshot has too many JSON nodes")

    def _scan_value(self, depth: int) -> None:
        if depth > ACCOUNT_SNAPSHOT_MAX_DEPTH:
            raise AccountSnapshotLimitError("account snapshot nesting is too deep")
        self._count_node()
        if self.index >= self.length:
            raise AccountSnapshotLimitError("truncated JSON value")
        byte = self.payload[self.index]
        if byte == 0x7B:
            self._scan_object(depth)
        elif byte == 0x5B:
            self._scan_array(depth)
        elif byte == 0x22:
            self._scan_string(ACCOUNT_SNAPSHOT_MAX_STRING_BYTES)
        elif byte == 0x74:
            self._scan_literal(b"true")
        elif byte == 0x66:
            self._scan_literal(b"false")
        elif byte == 0x6E:
            self._scan_literal(b"null")
        elif byte == 0x2D or 0x30 <= byte <= 0x39:
            self._scan_number()
        else:
            raise AccountSnapshotLimitError("invalid JSON value")

    def _scan_object(self, depth: int) -> None:
        self.index += 1
        self._skip_whitespace()
        if self.index < self.length and self.payload[self.index] == 0x7D:
            self.index += 1
            return
        fields = 0
        while True:
            fields += 1
            if fields > ACCOUNT_SNAPSHOT_MAX_OBJECT_FIELDS:
                raise AccountSnapshotLimitError("account snapshot object has too many fields")
            self._scan_string(ACCOUNT_SNAPSHOT_MAX_KEY_BYTES)
            self._skip_whitespace()
            self._consume(0x3A)
            self._skip_whitespace()
            self._scan_value(depth + 1)
            self._skip_whitespace()
            if self.index >= self.length:
                raise AccountSnapshotLimitError("truncated JSON object")
            delimiter = self.payload[self.index]
            self.index += 1
            if delimiter == 0x7D:
                return
            if delimiter != 0x2C:
                raise AccountSnapshotLimitError("invalid JSON object delimiter")
            self._skip_whitespace()

    def _scan_array(self, depth: int) -> None:
        self.index += 1
        self._skip_whitespace()
        if self.index < self.length and self.payload[self.index] == 0x5D:
            self.index += 1
            return
        items = 0
        while True:
            items += 1
            if items > ACCOUNT_SNAPSHOT_MAX_RECORDS:
                raise AccountSnapshotLimitError("account snapshot array has too many items")
            self._scan_value(depth + 1)
            self._skip_whitespace()
            if self.index >= self.length:
                raise AccountSnapshotLimitError("truncated JSON array")
            delimiter = self.payload[self.index]
            self.index += 1
            if delimiter == 0x5D:
                return
            if delimiter != 0x2C:
                raise AccountSnapshotLimitError("invalid JSON array delimiter")
            self._skip_whitespace()

    @staticmethod
    def _hex_value(byte: int) -> int:
        if 0x30 <= byte <= 0x39:
            return byte - 0x30
        if 0x41 <= byte <= 0x46:
            return byte - 0x41 + 10
        if 0x61 <= byte <= 0x66:
            return byte - 0x61 + 10
        raise AccountSnapshotLimitError("invalid JSON unicode escape")

    def _scan_u_escape(self) -> int:
        if self.index + 4 > self.length:
            raise AccountSnapshotLimitError("truncated JSON unicode escape")
        value = 0
        for byte in self.payload[self.index : self.index + 4]:
            value = (value << 4) | self._hex_value(byte)
        self.index += 4
        return value

    def _scan_string(self, max_bytes: int) -> None:
        self._consume(0x22)
        decoded_bytes = 0
        while self.index < self.length:
            byte = self.payload[self.index]
            self.index += 1
            if byte == 0x22:
                return
            if byte < 0x20:
                raise AccountSnapshotLimitError("unescaped control byte in JSON string")
            if byte == 0x5C:
                if self.index >= self.length:
                    raise AccountSnapshotLimitError("truncated JSON escape")
                escape = self.payload[self.index]
                self.index += 1
                if escape in b'"\\/bfnrt':
                    decoded_bytes += 1
                elif escape == 0x75:
                    codepoint = self._scan_u_escape()
                    if 0xD800 <= codepoint <= 0xDBFF:
                        if (
                            self.index + 2 > self.length
                            or self.payload[self.index] != 0x5C
                            or self.payload[self.index + 1] != 0x75
                        ):
                            raise AccountSnapshotLimitError("unpaired JSON surrogate")
                        self.index += 2
                        low = self._scan_u_escape()
                        if not 0xDC00 <= low <= 0xDFFF:
                            raise AccountSnapshotLimitError("unpaired JSON surrogate")
                        decoded_bytes += 4
                    elif 0xDC00 <= codepoint <= 0xDFFF:
                        raise AccountSnapshotLimitError("unpaired JSON surrogate")
                    elif codepoint <= 0x7F:
                        decoded_bytes += 1
                    elif codepoint <= 0x7FF:
                        decoded_bytes += 2
                    else:
                        decoded_bytes += 3
                else:
                    raise AccountSnapshotLimitError("invalid JSON escape")
            elif byte < 0x80:
                decoded_bytes += 1
            else:
                width = self._utf8_width(byte)
                end = self.index + width - 1
                if end > self.length:
                    raise AccountSnapshotLimitError("truncated UTF-8 sequence")
                sequence = self.payload[self.index - 1 : end]
                try:
                    sequence.decode("utf-8")
                except UnicodeDecodeError as exc:
                    raise AccountSnapshotLimitError("invalid UTF-8 sequence") from exc
                self.index = end
                decoded_bytes += width
            if decoded_bytes > max_bytes:
                raise AccountSnapshotLimitError("account snapshot string is too large")
        raise AccountSnapshotLimitError("truncated JSON string")

    @staticmethod
    def _utf8_width(lead: int) -> int:
        if 0xC2 <= lead <= 0xDF:
            return 2
        if 0xE0 <= lead <= 0xEF:
            return 3
        if 0xF0 <= lead <= 0xF4:
            return 4
        raise AccountSnapshotLimitError("invalid UTF-8 lead byte")

    def _scan_literal(self, literal: bytes) -> None:
        if self.payload[self.index : self.index + len(literal)] != literal:
            raise AccountSnapshotLimitError("invalid JSON literal")
        self.index += len(literal)

    def _scan_number(self) -> None:
        payload = self.payload
        length = self.length
        index = self.index
        if payload[index] == 0x2D:
            index += 1
            if index >= length:
                raise AccountSnapshotLimitError("truncated JSON number")
        if payload[index] == 0x30:
            index += 1
        elif 0x31 <= payload[index] <= 0x39:
            index += 1
            while index < length and 0x30 <= payload[index] <= 0x39:
                index += 1
        else:
            raise AccountSnapshotLimitError("invalid JSON number")
        if index < length and payload[index] == 0x2E:
            index += 1
            start = index
            while index < length and 0x30 <= payload[index] <= 0x39:
                index += 1
            if index == start:
                raise AccountSnapshotLimitError("invalid JSON fraction")
        if index < length and payload[index] in (0x65, 0x45):
            index += 1
            if index < length and payload[index] in (0x2B, 0x2D):
                index += 1
            start = index
            while index < length and 0x30 <= payload[index] <= 0x39:
                index += 1
            if index == start:
                raise AccountSnapshotLimitError("invalid JSON exponent")
        self.index = index


def validate_account_snapshot_bytes(payload: bytes) -> None:
    if not isinstance(payload, bytes) or len(payload) > ACCOUNT_SNAPSHOT_MAX_BYTES:
        raise AccountSnapshotLimitError("account snapshot byte limit exceeded")
    _JsonBudgetScanner(payload).scan()


def validate_account_records(value: object) -> list[dict[str, Any]]:
    if not isinstance(value, list) or len(value) > ACCOUNT_SNAPSHOT_MAX_RECORDS:
        raise AccountSnapshotLimitError("account snapshot record limit exceeded")
    stack: list[tuple[object, int]] = [(value, 1)]
    nodes = 0
    while stack:
        current, depth = stack.pop()
        if depth > ACCOUNT_SNAPSHOT_MAX_DEPTH:
            raise AccountSnapshotLimitError("account snapshot nesting is too deep")
        nodes += 1
        if nodes > ACCOUNT_SNAPSHOT_MAX_NODES:
            raise AccountSnapshotLimitError("account snapshot has too many JSON nodes")
        if isinstance(current, dict):
            if len(current) > ACCOUNT_SNAPSHOT_MAX_OBJECT_FIELDS:
                raise AccountSnapshotLimitError("account snapshot object has too many fields")
            for key, child in current.items():
                if not isinstance(key, str) or len(key.encode("utf-8")) > ACCOUNT_SNAPSHOT_MAX_KEY_BYTES:
                    raise AccountSnapshotLimitError("account snapshot key is invalid")
                stack.append((child, depth + 1))
        elif isinstance(current, list):
            if len(current) > ACCOUNT_SNAPSHOT_MAX_RECORDS:
                raise AccountSnapshotLimitError("account snapshot array has too many items")
            stack.extend((child, depth + 1) for child in current)
        elif isinstance(current, str):
            if len(current.encode("utf-8")) > ACCOUNT_SNAPSHOT_MAX_STRING_BYTES:
                raise AccountSnapshotLimitError("account snapshot string is too large")
        elif current is None or isinstance(current, (bool, int)):
            continue
        elif isinstance(current, float) and math.isfinite(current):
            continue
        else:
            raise AccountSnapshotLimitError("account snapshot contains a non-JSON value")
    if any(not isinstance(item, dict) for item in value):
        raise AccountSnapshotLimitError("account snapshot records must be objects")
    return cast(list[dict[str, Any]], value)

File: rust/src/account_snapshot.rs
use std::sync::LazyLock;

use serde::Deserialize;

pub(super) const MAX_BYTES: u64 = 16 * 1024 * 1024;
pub(super) const MAX_RECORDS: usize = 10_000;
pub(super) const MAX_DEPTH: usize = 16;
pub(super) const MAX_NODES: usize = 262_144;
pub(super) const MAX_OBJECT_FIELDS: usize = 128;
pub(super) const MAX_KEY_BYTES: usize = 256;
pub(super) const MAX_STRING_BYTES: usize = 16 * 1024;

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Contract {
    max_bytes: u64,
    max_records: usize,
    max_depth: usize,
    max_nodes: usize,
    max_object_fields: usize,
    max_key_bytes: usize,
    max_string_bytes: usize,
}

static CONTRACT_VALID: LazyLock<bool> = LazyLock::new(|| {
    let Ok(contract) =
        serde_json::from_str::<Contract>(include_str!("../../account_snapshot_contract.json"))
    else {
        return false;
    };
    contract.max_bytes == MAX_BYTES
        && contract.max_records == MAX_RECORDS
        && contract.max_depth == MAX_DEPTH
        && contract.max_nodes == MAX_NODES
        && contract.max_object_fields == MAX_OBJECT_FIELDS
        && contract.max_key_bytes == MAX_KEY_BYTES
        && contract.max_string_bytes == MAX_STRING_BYTES
});

pub(super) fn validate_bytes(payload: &[u8]) -> Result<(), ()> {
    if !*CONTRACT_VALID || payload.len() as u64 > MAX_BYTES {
        return Err(());
    }
    Scanner {
        payload,
        index: 0,
        nodes: 0,
    }
    .scan()
}

struct Scanner<'a> {
    payload: &'a [u8],
    index: usize,
    nodes: usize,
}

impl Scanner<'_> {
    fn scan(mut self) -> Result<(), ()> {
        self.skip_whitespace();
        self.scan_value(1)?;
        self.skip_whitespace();
        (self.index == self.payload.len()).then_some(()).ok_or(())
    }

    fn skip_whitespace(&mut self) {
        while self
            .payload
            .get(self.index)
            .is_some_and(|byte| matches!(byte, b' ' | b'\t' | b'\n' | b'\r'))
        {
            self.index += 1;
        }
    }

    fn consume(&mut self, expected: u8) -> Result<(), ()> {
        if self.payload.get(self.index).copied() != Some(expected) {
            return Err(());
        }
        self.index += 1;
        Ok(())
    }

    fn count_node(&mut self) -> Result<(), ()> {
        self.nodes = self.nodes.saturating_add(1);
        (self.nodes <= MAX_NODES).then_some(()).ok_or(())
    }

    fn scan_value(&mut self, depth: usize) -> Result<(), ()> {
        if depth > MAX_DEPTH {
            return Err(());
        }
        self.count_node()?;
        match self.payload.get(self.index).copied().ok_or(())? {
            b'{' => self.scan_object(depth),
            b'[' => self.scan_array(depth),
            b'"' => self.scan_string(MAX_STRING_BYTES),
            b't' => self.scan_literal(b"true"),
            b'f' => self.scan_literal(b"false"),
            b'n' => self.scan_literal(b"null"),
            b'-' | b'0'..=b'9' => self.scan_number(),
            _ => Err(()),
        }
    }

    fn scan_object(&mut self, depth: usize) -> Result<(), ()> {
        self.index += 1;
        self.skip_whitespace();
        if self.payload.get(self.index).copied() == Some(b'}') {
            self.index += 1;
            return Ok(());
        }
        let mut fields = 0usize;
        loop {
            fields = fields.saturating_add(1);
            if fields > MAX_OBJECT_FIELDS {
                return Err(());
            }
            self.scan_string(MAX_KEY_BYTES)?;
            self.skip_whitespace();
            self.consume(b':')?;
            self.skip_whitespace();
            self.scan_value(depth + 1)?;
            self.skip_whitespace();
            match self.payload.get(self.index).copied().ok_or(())? {
                b'}' => {
                    self.index += 1;
                    return Ok(());
                }
                b',' => {
                    self.index += 1;
                    self.skip_whitespace();
                }
                _ => return Err(()),
            }
        }
    }

    fn scan_array(&mut self, depth: usize) -> Result<(), ()> {
        self.index += 1;
        self.skip_whitespace();
        if self.payload.get(self.index).copied() == Some(b']') {
            self.index += 1;
            return Ok(());
        }
        let mut items = 0usize;
        loop {
            items = items.saturating_add(1);
            if items > MAX_RECORDS {
                return Err(());
            }
            self.scan_value(depth + 1)?;
            self.skip_whitespace();
            match self.payload.get(self.index).copied().ok_or(())? {
                b']' => {
                    self.index += 1;
                    return Ok(());
                }
                b',' => {
                    self.index += 1;
                    self.skip_whitespace();
                }
                _ => return Err(()),
            }
        }
    }

    fn scan_string(&mut self, max_bytes: usize) -> Result<(), ()> {
        self.consume(b'"')?;
        let mut decoded_bytes = 0usize;
        while let Some(byte) = self.payload.get(self.index).copied() {
            self.index += 1;
            match byte {
                b'"' => return Ok(()),
                0x00..=0x1f => return Err(()),
                b'\\' => {
                    let escape = self.payload.get(self.index).copied().ok_or(())?;
                    self.index += 1;
                    match escape {
                        b'"' | b'\\' | b'/' | b'b' | b'f' | b'n' | b'r' | b't' => {
                            decoded_bytes = decoded_bytes.saturating_add(1);
                        }
                        b'u' => {
                            let codepoint = self.scan_u_escape()?;
                            if (0xd800..=0xdbff).contains(&codepoint) {
                                if self.payload.get(self.index..self.index + 2) != Some(b"\\u") {
                                    return Err(());
                                }
                                self.index += 2;
                                let low = self.scan_u_escape()?;
                                if !(0xdc00..=0xdfff).contains(&low) {
                                    return Err(());
                                }
                                decoded_bytes = decoded_bytes.saturating_add(4);
                            } else if (0xdc00..=0xdfff).contains(&codepoint) {
                                return Err(());
                            } else {
                                decoded_bytes = decoded_bytes.saturating_add(match codepoint {
                                    0x0000..=0x007f => 1,
                                    0x0080..=0x07ff => 2,
                                    _ => 3,
                                });
                            }
                        }
                        _ => return Err(()),
                    }
                }
                0x20..=0x7f => decoded_bytes = decoded_bytes.saturating_add(1),
                lead => {
                    let width = utf8_width(lead).ok_or(())?;
                    let start = self.index - 1;
                    let end = start.checked_add(width).ok_or(())?;
                    let sequence = self.payload.get(start..end).ok_or(())?;
                    std::str::from_utf8(sequence).map_err(|_| ())?;
                    self.index = end;
                    decoded_bytes = decoded_bytes.saturating_add(width);
                }
            }
            if decoded_bytes > max_bytes {
                return Err(());
            }
        }
        Err(())
    }

    fn scan_u_escape(&mut self) -> Result<u16, ()> {
        let bytes = self.payload.get(self.index..self.index + 4).ok_or(())?;
        let mut value = 0u16;
        for byte in bytes {
            value = (value << 4) | u16::from(hex_value(*byte).ok_or(())?);
        }
        self.index += 4;
        Ok(value)
    }

    fn scan_literal(&mut self, literal: &[u8]) -> Result<(), ()> {
        if self.payload.get(self.index..self.index + literal.len()) != Some(literal) {
            return Err(());
        }
        self.index += literal.len();
        Ok(())
    }

    fn scan_number(&mut self) -> Result<(), ()> {
        if self.payload.get(self.index).copied() == Some(b'-') {
            self.index += 1;
        }
        match self.payload.get(self.index).copied().ok_or(())? {
            b'0' => self.index += 1,
            b'1'..=b'9' => {
                self.index += 1;
                while self.payload.get(self.index).is_some_and(u8::is_ascii_digit) {
                    self.index += 1;
                }
            }
            _ => return Err(()),
        }
        if self.payload.get(self.index).copied() == Some(b'.') {
            self.index += 1;
            let start = self.index;
            while self.payload.get(self.index).is_some_and(u8::is_ascii_digit) {
                self.index += 1;
            }
            if self.index == start {
                return Err(());
            }
        }
        if self
            .payload
            .get(self.index)
            .is_some_and(|byte| matches!(byte, b'e' | b'E'))
        {
            self.index += 1;
            if self
                .payload
                .get(self.index)
                .is_some_and(|byte| matches!(byte, b'+' | b'-'))
            {
                self.index += 1;
            }
            let start = self.index;
            while self.payload.get(self.index).is_some_and(u8::is_ascii_digit) {
                self.index += 1;
            }
            if self.index == start {
                return Err(());
            }
        }
        Ok(())
    }
}

fn hex_value(byte: u8) -> Option<u8> {
    match byte {
        b'0'..=b'9' => Some(byte - b'0'),
        b'a'..=b'f' => Some(byte - b'a' + 10),
        b'A'..=b'F' => Some(byte - b'A' + 10),
        _ => None,
    }
}

fn utf8_width(lead: u8) -> Option<usize> {
    match lead {
        0xc2..=0xdf => Some(2),
        0xe0..=0xef => Some(3),
        0xf0..=0xf4 => Some(4),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn array_with_items(count: usize) -> Vec<u8> {
        let mut payload = Vec::with_capacity(count.saturating_mul(2).saturating_add(2));
        payload.push(b'[');
        for index in 0..count {
            if index > 0 {
                payload.push(b',');
            }
            payload.push(b'0');
        }
        payload.push(b']');
        payload
    }

    fn object_with_fields(count: usize) -> Vec<u8> {
        let mut payload = Vec::new();
        payload.push(b'{');
        for index in 0..count {
            if index > 0 {
                payload.push(b',');
            }
            payload.extend_from_slice(format!(r#""k{index}":0"#).as_bytes());
        }
        payload.push(b'}');
        payload
    }

    fn nested_array_at_depth(depth: usize) -> Vec<u8> {
        let containers = depth.saturating_sub(1);
        let mut payload = vec![b'['; containers];
        payload.push(b'0');
        payload.extend(std::iter::repeat_n(b']', containers));
        payload
    }

    #[test]
    fn executable_contract_enforces_depth_array_and_object_boundaries() {
        assert!(validate_bytes(&nested_array_at_depth(MAX_DEPTH)).is_ok());
        assert!(validate_bytes(&nested_array_at_depth(MAX_DEPTH + 1)).is_err());
        assert!(validate_bytes(&array_with_items(MAX_RECORDS)).is_ok());
        assert!(validate_bytes(&array_with_items(MAX_RECORDS + 1)).is_err());
        assert!(validate_bytes(&object_with_fields(MAX_OBJECT_FIELDS)).is_ok());
        assert!(validate_bytes(&object_with_fields(MAX_OBJECT_FIELDS + 1)).is_err());
    }

    #[test]
    fn executable_contract_enforces_key_string_and_node_boundaries() {
        let key_at_limit = format!(r#"{{"{}":0}}"#, "k".repeat(MAX_KEY_BYTES));
        let key_over_limit = format!(r#"{{"{}":0}}"#, "k".repeat(MAX_KEY_BYTES + 1));
        assert!(validate_bytes(key_at_limit.as_bytes()).is_ok());
        assert!(validate_bytes(key_over_limit.as_bytes()).is_err());

        let string_at_limit = format!(r#""{}""#, "s".repeat(MAX_STRING_BYTES));
        let string_over_limit = format!(r#""{}""#, "s".repeat(MAX_STRING_BYTES + 1));
        assert!(validate_bytes(string_at_limit.as_bytes()).is_ok());
        assert!(validate_bytes(string_over_limit.as_bytes()).is_err());

        let mut node_bomb = Vec::new();
        node_bomb.push(b'[');
        for group in 0..MAX_RECORDS {
            if group > 0 {
                node_bomb.push(b',');
            }
            node_bomb.extend_from_slice(&array_with_items(26));
        }
        node_bomb.push(b']');
        assert!(node_bomb.len() as u64 <= MAX_BYTES);
        assert!(validate_bytes(&node_bomb).is_err());
    }
}
